### ISSUE TITLE
How do you use this code?

### ISSUE THREAD

**Issue by MrYossu** [UpwardInfo/BlazorPay#1]
Hello,

Thanks for posting up this code. I'm also struggling how to use Square with Blazor. Your code looks promising, but I'm not sure how I'd use it.

Please can you give some more details?

Thanks again.

**Comment by alfred-zaki** [comment]
this is a Blazor component so once you added it to your solution and reference it, add a using you should be able to use the tag <SquareCardPayment> 
parameters you need to supply are:
App Id: your Square App Id
App Location: an app location even if you have a default one you need to supply it's Id

How this works?
1- you show this component on your front end, which will collect card details and Toknize it.
2- you send the tokenized card details along with which transaction (invoice / order) this is for.
3- your back-end should get calc/fetch amount to be charged and use Square .NET SDK to "CreatePaymentRequest"
```csharp
                Square.SquareClient squareClient = new Square.SquareClient.Builder()
                    .Environment(Square.Environment.Sandbox)
                    .AccessToken(squareToken)
                    .Build();

                var amountMoney = new SquModels.Money.Builder()
                      .Amount(Convert.ToInt64(orderTotal * 100)) //Square amount is in cents
                      .Currency("USD")
                      .Build();

                var appFeeMoney = new SquModels.Money.Builder()
                  .Amount(100L)
                  .Currency("USD")
                  .Build();

                var body = new SquModels.CreatePaymentRequest.Builder(
                    sourceId: token, //token you got from front-end
                    idempotencyKey: Guid.NewGuid().ToString(),
                    amountMoney: amountMoney)
                  .AppFeeMoney(appFeeMoney)
                  .Autocomplete(true)
                  .ReferenceId(orderToPay.Id.ToString("N"))
                  .Note("it's good to add some reference to your transaction here")
                  .Build();


                var result = await squareClient.PaymentsApi.CreatePaymentAsync(body: body);
                
```

4- Check results and send your front end success or failure.

hope this helps

**Comment by MrYossu** [comment]
@alfred-zaki Thanks for the quick reply. I wish Square's support replied as quickly as you did!

How do I tell it that I want to use the sandbox? I tried using the code as you said, but got an exception "_Microsoft.JSInterop.JSException: Web Payments SDK was initialized with an application ID created in sandbox however you are currently using production_"

I can't see anywhere in your component that you can tell it whether to use sandbox or production.

Thanks again.

**Comment by MrYossu** [comment]
@alfred-zaki Sorry, another question. The JavaScript uses v1 of the API. Do you know if the code will work with v2? Might as well use the latest.

Thanks again.

**Comment by MrYossu** [comment]
@alfred-zaki What is `SquModels`? Is that your code, or something in their SDK? Can't find it anywhere.

Thanks

**Comment by MrYossu** [comment]
In case it helps anyone, I had to change the code slightly to get it to work. Two points worth noting...

1) In the JS file, the URL shown (`https://web.squarecdn.com/v1/square.js`) is for production. If you want to test in the sandbox, you need to use `https://sandbox.web.squarecdn.com/v1/square.js`

2) The references to `SquModels` in the code above didn't resolve, but by removing that part, it worked fine. So, you change...

`var amountMoney = new SquModels.Money.Builder()`

...to...

`var amountMoney = new Money.Builder()`

...and it all works fine!

Thanks to @alfred-zaki for creating the component and posting the code. I'd still like to know if we can use v2 of the API. Just changing the reference in the two URLs shown above gave a 404, so I guess we'd need to find the URLs of the v2 JS and then test to see if the code shown works.

**Comment by alfred-zaki** [comment]
@MrYossu glad that you figured things out, please let me know if you need any more help from my side.

**Comment by alfred-zaki** [comment]
@MrYossu looks good to me 👍

**Comment by MrYossu** [comment]
@alfred-zaki Thanks very much for all the help!

**Comment by alfred-zaki** [comment]
@nssidhu though you might be interested!

**Comment by MrYossu** [comment]
@alfred-zaki Me again. Have you worked out how to set up a recurring payment with Square? As you can see from my sample, I can create an order and take payment for it, but that's as a one-off. I need to be able to set up recurring payments as well.

Have you done this?

Thanks again.

**Comment by MrYossu** [comment]
@alfred-zaki Duh, I was looking at the card API! Didn't notice the subscriptions one.

Thanks

**Comment by MrYossu** [comment]
@alfred-zaki OK, so I'm stuck! If you grab my [latest sample code](https://github.com/MrYossu/SquarePayments), you'll see I updated the customers page so that when you create a customer, it tries to create a card as well (which is needed for a subscription where the payment is taken automatically). Please see lines 75 and onwards in `Pages/Customer.razor.cs`.

Trouble is, I get an error when trying to create the card. It's pretty unhelpful, just tells me that there was an error with the source Id, which I think is the payment token. I'm using the same code as you originally showed me to get a payment token, but it doesn't work.

As you can see from the code, I tried using the sandbox source Id of `cnon:card-nonce-ok` but that also results in an error.

Any ideas? Thanks again.

**Comment by attiqdev** [comment]
Hi Guys, thanks for this code sample I am using your sample @MrYossu but I am facing one issue can you please help me with this. When I am calling the  CreateOrderResponse orderResponse = await _client.OrdersApi.CreateOrderAsync(orderRequest); in SquareHelper.cs line 139, I am getting this error 

"Access to fetch at 'https://connect.squareupsandbox.com/v2/orders' from origin 'https://localhost:44309' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: It does not have HTTP ok status."

can you please tell me what else should I do to reslove this?

**Comment by MrYossu** [comment]
@attiqdev Does this stop the code working, or do you just get those warnings in the console panel?

The reason I ask is that I have seen things like this in the console, but it's never stopped my code from working. If your code isn't working because of this, then there's something else going on.

**Comment by attiqdev** [comment]
Actually @MrYossu , I am using this code in the blazer application the same as you are using but when I reach this line where I create and get the order in the helper class it's giving me the exception for CORS. So I am wondering how its working for you with doing anything for CORS

**Comment by MrYossu** [comment]
@attiqdev I never had any problems with it. I just tried it again, and used https://localhost:7278/OrderAndPay to place an order. All went through fine.

Sorry, I don't know what to tell you. In case it helps, I'll paste in the latest version of the helper I have. This has changed a bit from what's in the example, as I was developing this in the app where it was used, and haven't had time to update this repo. The API has changed slightly, but you should be able to work out what to do. The main changes here were to try and wrap as much Square code in the helper as possible, which means not exposing the Square client or location Id. I left those in the class (at the bottom), as I still have some older code that I haven't changed yet.

Don't know if it will help, as the repo code works for me, but might be worth a try.

```c#
using Microsoft.JSInterop;
using Square;
using Square.Exceptions;
using Square.Models;

namespace MyProject.Helpers;

public class SquareHelper {
  #region Variables

  private readonly IJSRuntime _js;
  private readonly SquareData _data;
  private readonly SquareClient _client;
  private IJSObjectReference? _squareJs;
  private IJSObjectReference? _squareCard;
  private string _elementId = "";
  public static string JsUri = "/Square.js";

  #endregion

  public SquareHelper(SquareData data, IJSRuntime js) {
    _data = data;
    _js = js;
    _client = new SquareClient.Builder()
      .Environment(_data.Environment == "production" ? Square.Environment.Production : Square.Environment.Sandbox)
      .AccessToken(_data.AccessToken)
      .Build();
  }

  #region Miscellaneous

  public static Address BuildAddress(string address1, string address2, string postcode, string country) {
    Address address = new Address.Builder()
      .AddressLine1(address1)
      .AddressLine2(address2)
      .PostalCode(postcode)
      .Country(country)
      .Build();
    return address;
  }

  public async Task<string> CreateSubscription(string customerId, string cardId, string planId, int dayOfMonth = 0) {
    DateTime start = dayOfMonth == 0 ? DateTime.Now : DateTime.Now.EndOfMonth().AddMilliseconds(2);
    CreateSubscriptionRequest subscriptionRequest = new CreateSubscriptionRequest.Builder(_data.LocationId, planId, customerId)
      .IdempotencyKey(Guid.NewGuid().ToString("N"))
      .CardId(cardId)
      .StartDate(start.ToString("yyyy-MM-dd"))
      .Build();
    CreateSubscriptionResponse subscriptionResponse = await _client.SubscriptionsApi.CreateSubscriptionAsync(subscriptionRequest);
    return subscriptionResponse.Subscription.Id;
  }

  #endregion

  #region Cards

  public async Task SetUpCard(string elementId) {
    _elementId = elementId;
    await _js.InvokeAsync<IJSObjectReference>("import", _data.Environment == "production" ? "https://web.squarecdn.com/v1/square.js" : "https://sandbox.web.squarecdn.com/v1/square.js");
    _squareJs = await _js.InvokeAsync<IJSObjectReference>("import", JsUri);
    _squareCard = await _squareJs.InvokeAsync<IJSObjectReference>("addSquareCardPayment", _elementId, _data.AppId, _data.LocationId);
  }

  public async Task<SquareResult<string>> ChargeCard(string cardId, long amount, string customerId, string note, object contact) {
    try {
      string verificationToken = await _squareJs.InvokeAsync<string>("verifyTheBuyer", _data.AppId, _data.LocationId, cardId, (amount / 100.0).ToString("F2"), contact);
      CreatePaymentRequest body = new CreatePaymentRequest.Builder(cardId, Guid.NewGuid().ToString(), new Money.Builder()
          .Amount(amount)
          .Currency("GBP")
          .Build())
        .Autocomplete(true)
        .CustomerId(customerId)
        .VerificationToken(verificationToken)
        .LocationId(_data.LocationId)
        .Note(note)
        .Build();
      CreatePaymentResponse result = await _client.PaymentsApi.CreatePaymentAsync(body: body);
      if (result.Errors?.Any() ?? false) {
        return BuildErrors(result.Errors);
      }
      return new(SquareStatus.Success, result.Payment.Id);
    } catch (ApiException ex) {
      return BuildErrors(ex.Errors);
    } catch (Exception ex) {
      if (ex.Message.ToLower().Contains("verif")) {
        return new(SquareStatus.FailedVerification, ex.Message);
      }
      return new(SquareStatus.Error, ex.Message);
    }
  }

  public async Task<SquareResult<string>> CreateCard(string firstName, string surname, Address address, string customerId, object contact) {
    try {
      string sourceId = await _squareJs.InvokeAsync<string>("getSquareCardToken", _squareCard);
      string verificationToken = await _squareJs.InvokeAsync<string>("verifyTheBuyerCard", _data.AppId, _data.LocationId, sourceId, contact);
      // NOTE - The following will throw an error if the postcode/zip used in the card element does not match the one you use in the address
      CreateCardRequest cardRequest = new CreateCardRequest.Builder(Guid.NewGuid().ToString(), sourceId,
          new Card.Builder()
            .CardholderName($"{firstName} {surname}")
            .BillingAddress(address)
            .CustomerId(customerId)
            .ReferenceId($"{firstName}{surname}")
            .Build())
        .VerificationToken(verificationToken)
        .Build();
      CreateCardResponse cardResponse = await _client.CardsApi.CreateCardAsync(cardRequest);
      if (cardResponse.Errors?.Any() ?? false) {
        return BuildErrors(cardResponse.Errors);
      }
      return new(SquareStatus.Success, cardResponse.Card.Id);
    }
    catch (ApiException ex) {
      return BuildErrors(ex.Errors);
    }
    catch (Exception ex) {
      if (ex.Message.ToLower().Contains("verif")) {
        return new(SquareStatus.FailedVerification, ex.Message);
      }
      return new(SquareStatus.Error, ex.Message);
    }
  }

  #endregion

  #region Payments

  public async Task<SquareResult<string>> TakePayment(string customerId, List<OrderLineItem> lineItems, string note, object contact) {
    CreateOrderRequest orderRequest = new CreateOrderRequest.Builder()
      .Order(new Order.Builder(_data.LocationId)
        .CustomerId(customerId)
        .LineItems(lineItems)
        .Build())
      .IdempotencyKey(Guid.NewGuid().ToString())
      .Build();
    CreateOrderResponse orderResponse = await _client.OrdersApi.CreateOrderAsync(orderRequest);
    string orderId = orderResponse.Order.Id;
    long total = lineItems.Sum(li => (li.BasePriceMoney.Amount ?? 0) * Convert.ToInt32(li.Quantity));
    return await GetCreatePaymentRequest(total, note, contact, orderId)
      .Match(async paymentRequest => await TakePaymentDo(paymentRequest), _ => new(SquareStatus.FailedVerification, ""));
  }

  public async Task<SquareResult<string>> TakePayment(long totalInPennies, string note, object contact) =>
    await GetCreatePaymentRequest(totalInPennies, note, contact)
      .Match(async paymentRequest => await TakePaymentDo(paymentRequest), _ => new(SquareStatus.FailedVerification, ""));

  private async Task<SquareResult<string>> TakePaymentDo(CreatePaymentRequest body) {
    try {
      CreatePaymentResponse result = await _client.PaymentsApi.CreatePaymentAsync(body);
      if (result.Errors?.Any() ?? false) {
        if (result.Errors.Any(e => e.Detail.Contains("GENERIC_DECLINE"))) {
          return BuildErrors(result.Errors);
        }
        return new(SquareStatus.Error, result.Errors.Select(e => $"({e.Field}) {e.Detail}").JoinStr());
      }
      return new(SquareStatus.Success, result.Payment.Id);
    }
    catch (ApiException ex) {
      return BuildErrors(ex.Errors);
    }
    catch (Exception ex) {
      return new(SquareStatus.Error, ex.Message);
    }
  }

  public static SquareResult<string> BuildErrors(IList<Error> errors) {
    if (errors.Any(e => e.Detail.Contains("GENERIC_DECLINE"))) {
      return new(SquareStatus.Error, "The payment request was declined by your bank");
    }
    if (errors.Any(e => e.Detail.ToLower().Contains("verif"))) {
      return new(SquareStatus.Error, "Your card could not be verified");
    }
    return new(SquareStatus.Error, errors.Select(e => $"({e.Field}) {e.Detail}").JoinStr());
  }

  private TryAsync<CreatePaymentRequest> GetCreatePaymentRequest(long totalInPennies, string note, object contact, string orderId = "") =>
    TryAsync(() => GetCreatePaymentRequestDo(totalInPennies, note, contact, orderId));

  private async Task<CreatePaymentRequest> GetCreatePaymentRequestDo(long totalInPennies, string note, object contact, string orderId = "") {
    string sourceId = await _squareJs.InvokeAsync<string>("getSquareCardToken", _squareCard);
    string verificationToken = await _squareJs.InvokeAsync<string>("verifyTheBuyer", _data.AppId, _data.LocationId, sourceId, (totalInPennies / 100.0).ToString("F2"), contact);
    CreatePaymentRequest.Builder body = new CreatePaymentRequest.Builder(
        sourceId, Guid.NewGuid().ToString(), new Money.Builder()
          .Amount(totalInPennies)
          .Currency("GBP")
          .Build())
      .Autocomplete(true)
      .VerificationToken(verificationToken)
      .Note(note);
    if (!string.IsNullOrWhiteSpace(orderId)) {
      body.OrderId(orderId);
    }
    return body.Build();
  }

  #endregion

  #region Customers

  public async Task<List<Customer>> GetAllCustomers() {
    List<Customer> results = new();
    ListCustomersResponse customerResponse = await _client.CustomersApi.ListCustomersAsync();
    results.AddRange(customerResponse.Customers);
    string cursor = customerResponse.Cursor;
    while (!string.IsNullOrWhiteSpace(cursor)) {
      customerResponse = await _client.CustomersApi.ListCustomersAsync(cursor);
      if (customerResponse.Customers == null) {
        break;
      }
      results.AddRange(customerResponse.Customers);
      cursor = customerResponse.Cursor;
    }
    return results;
  }

  public async Task<List<Customer>> SearchCustomers(string email, bool exact = true) {
    CustomerTextFilter filter = exact
      ? new CustomerTextFilter.Builder()
        .Exact(email)
        .Build()
      : new CustomerTextFilter.Builder()
        .Fuzzy(email)
        .Build();
    SearchCustomersRequest.Builder body = new SearchCustomersRequest.Builder()
      .Query(new CustomerQuery.Builder()
        .Filter(new CustomerFilter.Builder()
          .EmailAddress(filter)
          .Build())
        .Build());
    List<Customer> results = new();
    try {
      SearchCustomersResponse customerResponse = await _client.CustomersApi.SearchCustomersAsync(body.Build());
      if (customerResponse.Customers == null) {
        return results;
      }
      results.AddRange(customerResponse.Customers);
      string cursor = customerResponse.Cursor;
      while (!string.IsNullOrWhiteSpace(cursor)) {
        customerResponse = await _client.CustomersApi.SearchCustomersAsync(body.Cursor(cursor).Build());
        if (customerResponse.Customers == null) {
          break;
        }
        results.AddRange(customerResponse.Customers);
        cursor = customerResponse.Cursor;
      }
      return results;
    }
    catch {
      return results;
    }
  }

  public async Task<string> CreateCustomer(string firstName, string surname, string email, string phone, Address address) {
    CreateCustomerRequest customerRequest = new CreateCustomerRequest.Builder()
      .IdempotencyKey(Guid.NewGuid().ToString())
      .GivenName(firstName)
      .FamilyName(surname)
      .EmailAddress(email)
      .Address(address)
      .PhoneNumber(phone)
      .Build();
    CreateCustomerResponse customerResponse = await _client.CustomersApi.CreateCustomerAsync(customerRequest);
    return customerResponse.Customer.Id;
  }

  public async Task<DeleteCustomerResponse> DeleteCustomer(string id) =>
    await _client.CustomersApi.DeleteCustomerAsync(id);

  #endregion

  #region Deprecated (move the calling code into this helper)

  public SquareClient GetSquareClient() =>
    _client;

  public string LocationId =>
    _data.LocationId;

  #endregion
}
```

**Comment by attiqdev** [comment]
Sure I will try this one thing I need to ask I am attaching the square card div with my page div and then adding the test card (from square) and when I click on pay I call this function **OrderWithPayment** from helper then I am getting the CORS issue.

![image](https://user-images.githubusercontent.com/103983141/170889676-ba52f938-133a-42ab-a001-1357bbd59ae4.png)

**Comment by MrYossu** [comment]
Hmm, not sure. Did you see [this SO answer](https://stackoverflow.com/a/58219953/706346)? Gives more info about the error, which might help.

If not, I guess you'll have to ask your own question on SO. I wish I could help more, but I haven't had this issue.

If/when you get it sorted, please come back and explain what you did, as it may help someone else.

Good luck!
